File: app/src/main/java/com/eje_c/reastream/ReaStream.java
package com.eje_c.reastream;

import android.annotation.SuppressLint;

import com.eje_c.libreastream.AudioRecordSrc;
import com.eje_c.libreastream.AudioTrackSink;
import com.eje_c.libreastream.ReaStreamPacket;
import com.eje_c.libreastream.ReaStreamReceiver;
import com.eje_c.libreastream.ReaStreamSender;

import java.io.IOException;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.FloatBuffer;
import java.util.Arrays;

public class ReaStream implements AutoCloseable {
    private boolean recording;
    private boolean playing;
    private int sampleRate = 44100;
    private ReaStreamSender sender;     // Non null while sending
    private ReaStreamReceiver receiver; // Non null while receiving
    private boolean enabled = true;
    private InetAddress remoteAddress;

    public void startSending() {

        if (!recording) {
            recording = true;
            new SenderThread().start();
        }
    }

    public void stopSending() {
        recording = false;
    }

    public void startReveiving() {

        if (!playing) {
            playing = true;
            new ReceiverThread().start();
        }
    }

    public void stopReceiving() {
        playing = false;
    }

    public boolean isSending() {
        return recording;
    }

    public boolean isReceiving() {
        return playing;
    }

    public void setIdentifier(String identifier) {

        if (sender != null) {
            sender.setIdentifier(identifier);
        }

        if (receiver != null) {
            receiver.setIdentifier(identifier);
        }
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public boolean isEnabled() {
        return enabled;
    }

    public void setRemoteAddress(String remoteAddress) throws UnknownHostException {
        setRemoteAddress(InetAddress.getByName(remoteAddress));
    }

    public void setRemoteAddress(InetAddress remoteAddress) {
        this.remoteAddress = remoteAddress;

        if (sender != null) {
            sender.setRemoteAddress(remoteAddress);
        }
    }

    @Override
    public void close() {

        if (sender != null) {
            sender.close();
        }

        if (receiver != null) {
            receiver.close();
        }
    }

    private class SenderThread extends Thread {

        @SuppressLint("NewApi")
        @Override
        public void run() {

            try (ReaStreamSender sender = new ReaStreamSender();
                 AudioRecordSrc audioRecordSrc = new AudioRecordSrc(sampleRate)) {

                audioRecordSrc.start();
                ReaStream.this.sender = sender;

                sender.setSampleRate(sampleRate);
                sender.setChannels((byte) 1);
                sender.setRemoteAddress(remoteAddress);

                while (recording) {

                    // Read from mic and send it
                    FloatBuffer buffer = audioRecordSrc.read();
                    int readCount = buffer.limit();
                    if (enabled && readCount > 0) {
                        sender.send(buffer.array(), readCount);
                    }
                }

                audioRecordSrc.stop();

            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                ReaStream.this.sender = null;
            }
        }
    }

    private class ReceiverThread extends Thread {

        @SuppressLint("NewApi")
        @Override
        public void run() {

            try (ReaStreamReceiver receiver = new ReaStreamReceiver();
                 AudioTrackSink audioTrackSink = new AudioTrackSink(sampleRate)) {
                audioTrackSink.start();
                ReaStream.this.receiver = receiver;

                while (playing) {
                    if (enabled) {
                        ReaStreamPacket packet = receiver.receive();
                        if (packet.isAudioData()) {
                            audioTrackSink.onReceive(packet);
                        } else if (packet.isMidiData()) {
                            // TODO
                            System.out.println(Arrays.toString(packet.midiEvents));
                        }
                    }
                }

                audioTrackSink.stop();

            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                ReaStream.this.receiver = null;
            }
        }
    }
}

File: app/src/main/java/com/eje_c/reastream/ReaStream.kt
package com.eje_c.reastream

import android.annotation.SuppressLint
import com.eje_c.libreastream.*
import java.io.IOException
import java.net.InetAddress
import java.net.UnknownHostException

class ReaStream(val sampleRate: Int = 44100) : AutoCloseable {

    var isSending: Boolean = false
        private set

    var isReceiving: Boolean = false
        private set

    var isEnabled = true

    var onMidiEvents: ((Array<MidiEvent>) -> Unit)? = null

    private var sender: ReaStreamSender? = null     // Non null while sending
    private var receiver: ReaStreamReceiver? = null // Non null while receiving
    private var remoteAddress: InetAddress? = null

    /**
     * Start sending audio from microphone.
     */
    fun startSending() {

        if (!isSending) {
            isSending = true
            SenderThread().start()
        }
    }

    /**
     * Stop sending audio from microphone.
     */
    fun stopSending() {
        isSending = false
    }

    /**
     * Start receiving audio.
     */
    fun startReveiving() {

        if (!isReceiving) {
            isReceiving = true
            ReceiverThread().start()
        }
    }

    /**
     * Stop receiving audio.
     */
    fun stopReceiving() {
        isReceiving = false
    }

    /**
     * Set identifier. Identifier is used to identify audio stream.
     */
    fun setIdentifier(identifier: String) {

        if (sender != null) {
            sender!!.identifier = identifier
        }

        if (receiver != null) {
            receiver!!.identifier = identifier
        }
    }

    @Throws(UnknownHostException::class)
    fun setRemoteAddress(remoteAddress: String) {
        setRemoteAddress(InetAddress.getByName(remoteAddress))
    }

    fun setRemoteAddress(remoteAddress: InetAddress) {
        this.remoteAddress = remoteAddress

        if (sender != null) {
            sender!!.remoteAddress = remoteAddress
        }
    }

    override fun close() {
        sender?.close()
        receiver?.close()
    }

    private inner class SenderThread : Thread() {

        override fun run() {

            try {
                ReaStreamSender().use { sender ->
                    AudioRecordSrc(sampleRate).use { audioRecordSrc ->

                        audioRecordSrc.start()
                        this@ReaStream.sender = sender

                        sender.sampleRate = sampleRate
                        sender.channels = 1.toByte()
                        sender.remoteAddress = remoteAddress!!

                        while (isSending) {

                            // Read from mic and send it
                            val buffer = audioRecordSrc.read()
                            val readCount = buffer.limit()
                            if (isEnabled && readCount > 0) {
                                sender.send(buffer.array(), readCount)
                            }
                        }

                        audioRecordSrc.stop()

                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                this@ReaStream.sender = null
            }
        }
    }

    private inner class ReceiverThread : Thread() {

        @SuppressLint("NewApi")
        override fun run() {

            try {
                ReaStreamReceiver().use { receiver ->
                    AudioTrackSink(sampleRate).use { audioTrackSink ->
                        this@ReaStream.receiver = receiver

                        while (isReceiving) {
                            if (isEnabled) {
                                val packet = receiver.receive()
                                if (packet.isAudioData) {
                                    audioTrackSink.onReceive(packet)
                                } else if (packet.isMidiData) {
                                    onMidiEvents?.invoke(packet.midiEvents!!)
                                }
                            }
                        }
                    }
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                this@ReaStream.receiver = null
            }
        }
    }
}
